Add support for multiple runtimes

mlaventure opened the pull request:
Add support for allowing selecting a different oci runtime at container creation.

Closes #22845

**Notes**: PR for engine-api repo will be done once this is approved.

mlaventure committed:
Add support for multiples runtimes

Signed-off-by: Kenfe-Mickael Laventure <mickael.laventure@gmail.com>

mlaventure committed:
Add missing completion for --config-file

Signed-off-by: Kenfe-Mickael Laventure <mickael.laventure@gmail.com>

icecrime commented on api/client/info.go:
```diff
@@ -68,6 +68,17 @@ func (cli *DockerCli) CmdInfo(args ...string) error {
 		fmt.Fprintf(cli.out, "\n")
 	}
 
+	if len(info.Runtimes) > 0 {
```
Can there really be none?

mlaventure replied:
No such concept on Windows.

It shows a
```
Runtime:
```
line in info without that check

icecrime commented on daemon/config_unix.go:
```diff
@@ -82,6 +85,37 @@ func (config *Config) InstallFlags(cmd *flag.FlagSet, usageFn func(string) strin
 	cmd.StringVar(&config.RemappedRoot, []string{"-userns-remap"}, "", usageFn("User/Group setting for user namespaces"))
 	cmd.StringVar(&config.ContainerdAddr, []string{"-containerd"}, "", usageFn("Path to containerd socket"))
+	config.Runtimes = make(map[string]runtime.Runtime)
+	cmd.Var(runconfigopts.NewNamedRuntimeOpt("runtimes", &config.Runtimes), []string{"-runtime"}, usageFn("Add an OCI runtime to the list of available runtimes"))
```
`"Register an additional OCI compatible runtime"`?

icecrime replied:
I wonder if we should name this `--add-runtime` to make it clearer.

mlaventure replied:
That was the initial name. ping @crosbymichael

crosbymichael replied:
its fine, whatever you think is best

WeiZhang555 replied:
I vote to `--add-runtime`, because containerd has a `--runtime` flag with different meaning :smile:

mlaventure replied:
Yes, I found that too this morning, fixing it and adding more tests. Thanks
:)

On Sun, May 29, 2016 at 8:52 PM zhangwei_cs <notifications@github.com>
wrote:

> In daemon/config_unix.go
> <https://github.com/docker/docker/pull/22983#discussion_r65021281>:
>
> > @@ -82,6 +85,37 @@ func (config *Config) InstallFlags(cmd *flag.FlagSet, usageFn func(string) strin
> >  	cmd.StringVar(&config.CgroupParent, []string{"-cgroup-parent"}, "", usageFn("Set parent cgroup for all containers"))
> >  	cmd.StringVar(&config.RemappedRoot, []string{"-userns-remap"}, "", usageFn("User/Group setting for user namespaces"))
> >  	cmd.StringVar(&config.ContainerdAddr, []string{"-containerd"}, "", usageFn("Path to containerd socket"))
> > +	config.Runtimes = make(map[string]runtime.Runtime)
> > +	cmd.Var(runconfigopts.NewNamedRuntimeOpt("runtimes", &config.Runtimes), []string{"-runtime"}, usageFn("Add an OCI runtime to the list of available runtimes"))
>
> I vote to --add-runtime, because containerd has a --runtime flag with
> different meaning 😄
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/docker/docker/pull/22983/files/da0548d8aa947a2cc61aada1a6490c4c87190a40..53662739e29cfbfd11231be27104226d9ba3d828#r65021281>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/APUiflXzX61bWfGbPBtUVhdu-PewrviLks5qGl7rgaJpZM4Imvu8>
> .
>

icecrime commented on daemon/config_unix.go:
```diff
@@ -82,6 +85,37 @@ func (config *Config) InstallFlags(cmd *flag.FlagSet, usageFn func(string) strin
 	cmd.StringVar(&config.ContainerdAddr, []string{"-containerd"}, "", usageFn("Path to containerd socket"))
+	config.Runtimes = make(map[string]runtime.Runtime)
+	cmd.Var(runconfigopts.NewNamedRuntimeOpt("runtimes", &config.Runtimes), []string{"-runtime"}, usageFn("Add an OCI runtime to the list of available runtimes"))
+	cmd.StringVar(&config.DefaultRuntime, []string{"-default-runtime"}, "", usageFn("Default OCI runtime to be used"))
```
I don't think this one should be a command-line flag: it should be enforced at compile-time.

icecrime commented on daemon/daemon.go:
```diff
@@ -66,6 +66,10 @@ import (
 var (
+	// DefaultRuntimePath is the default runtime to be used by
+	// containerd if none is specified
+	DefaultRuntimePath = "docker-runc"
```
It's not really a path ;-) Can you rename to `DefaultRuntimeBinary`?

mlaventure replied:
Sure

icecrime commented:
LGTM on the API changes. A few nits on the daemon integration.

Cool :+1:

WeiZhang555 commented on runconfig/opts/parse.go:
```diff
@@ -100,6 +100,7 @@ func Parse(cmd *flag.FlagSet, args []string) (*container.Config, *container.Host
 		flStopSignal        = cmd.String([]string{"-stop-signal"}, signal.DefaultStopSignal, fmt.Sprintf("Signal to stop a container, %v by default", signal.DefaultStopSignal))
 		flIsolation         = cmd.String([]string{"-isolation"}, "", "Container isolation technology")
 		flShmSize           = cmd.String([]string{"-shm-size"}, "", "Size of /dev/shm, default value is 64MB")
+		flRuntime           = cmd.String([]string{"-runtime"}, "", "Runtime to use with the container")
```
Should we support a short command `-r`? So it can be consistent with `containerd`

thaJeztah replied:
We try to not add shorthand flags, unless it's a really much used option, so I'd prefer to not add it. We can always add in future; removing such options is more troublesome than adding

thaJeztah replied:
Same probably applies for containerd - do we really need a shorthand option? @mlaventure ?

mlaventure replied:
No we don't, we can always remove it.

WeiZhang555 replied:
Yay, I see, it makes sense :)

WeiZhang555 commented:
I've tried this and found that it didn't respect default runtime value from config. 

`/etc/docker/daemon.json`:
```
{
    "default-runtime": "foo",
    "runtimes": {
        "oci": {
            "path": "/usr/local/bin/docker-runc",
            "runtimeArgs": [
                "--debug"
            ]   
        },  
        "foo": {
            "path": "/usr/local/bin/runc",
            "runtimeArgs": [
                "--debug"
            ]   
        }   
    }   
}
```

restart docker daemon and show docker info:
```
$ dockerd -D -s aufs
$ docker info
Runtimes:
 foo: /usr/local/bin/runc (--debug)
 oci: /usr/local/bin/docker-runc (--debug)

<default is missing...> 
<send SIGHUP to trigger daemon reload>

$ docker info
Runtimes:
 default: docker-runc
 foo: /usr/local/bin/runc (--debug)
 oci: /usr/local/bin/docker-runc (--debug)

<default is expected to be foo>
```

Correct me if I miss something :-)

WeiZhang555 commented:
And another unexpected bahavior:
```
$ docker info
Runtimes:
 default: docker-runc
 foo: /usr/local/bin/runc (--debug)
 oci: /usr/local/bin/docker-runc (--debug)

$ docker run -tid --runtime foo busybox sh
b888e333818c37436811b0835bc1697c923a95772e23b72af24b933441bb565b
$ docker inspect -f "{{.HostConfig.Runtime}}" b888e333818c

<it's all correct so far>
<modify value of `foo` and let daemon reload>

$ docker info
Runtimes:
 default: docker-runc
 foo: /usr/local/bin/foo (--debug)
 oci: /usr/local/bin/docker-runc (--debug)

$ docker run -tid --runtime foo busybox sh
295267a6ea7e1c5b2fc4c3bfebdbd00ce8be52dda0aa602566ef49e792d8e247
$ docker inspect -f "{{.HostConfig.Runtime}}" 295267a6ea7e
foo
$ docker ps 
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
295267a6ea7e        busybox             "sh"                47 seconds ago      Up 45 seconds                           serene_fermi
b888e333818c        busybox             "sh"                7 minutes ago       Up 7 minutes                            awesome_goodall

< /usr/local/bin/foo doesn't exist, so 295267a6ea7e isn't expected to be running...>
```

mlaventure commented:
Also, "default" is going to be "docker-runc" all the time that's the expected behavior. I should add a "Default Runtime" field to the output of `docker info` I guess

mlaventure commented:
@WeiZhang555 I couldn't reproduce your second issue. I've updated the code, to take care of some other cases that were missing. Please try again :)

WeiZhang555 commented:
@mlaventure Sorry, it's my bad. I'm using official dockercore/docker to compile this piece of code and forget that containerd isn't updated. After compiling latest version of containerd, it works perfectly.
I'll try to play more with this to see if I can find something else :stuck_out_tongue_winking_eye:

crosbymichael commented:
LGTM

vdemeester commented:
@mlaventure you can make a PR on engine-api now :angel:

mlaventure commented:
Waiting on https://github.com/docker/docker/pull/23090 to get merged before vendoring engine-api. The build fails otherwise.

thaJeztah commented:
@mlaventure https://github.com/docker/docker/pull/23090 was just merged, so you can rebase

cpuguy83 commented:
Needs a rebase and docs

mlaventure committed:
Add bash completion support for --runtime and --add-runtime

Signed-off-by: Kenfe-Mickael Laventure <mickael.laventure@gmail.com>

SvenDowideit commented:
retest this please

mlaventure commented:
- Rebased
- Added docs

thaJeztah commented:
ping @cpuguy83 @vdemeester PTAL if this is ready to go to docs review

cpuguy83 commented on daemon/config_windows.go:
```diff
@@ -40,3 +41,19 @@ func (config *Config) InstallFlags(cmd *flag.FlagSet, usageFn func(string) strin
+
+// GetDefaultRuntimeName returns the current default runtime
+func (config *Config) GetDefaultRuntimeName() string {
+	return types.DefaultRuntimeName
```
Curious why we're putting this in engine-api.

mlaventure replied:
uh, when I moved the `Runtime` type to the API, this moved with it.

I'll make a subsequent PR to move `DefaultRuntimeName`. Although not sure where I'll put it in order to avoid a circular dependency between the `daemon` package and the `opts` one.

cpuguy83 commented on daemon/daemon.go:
```diff
@@ -785,10 +789,24 @@ func (daemon *Daemon) initDiscovery(config *Config) error {
 func (daemon *Daemon) Reload(config *Config) error {
+	var err error
+	// used to hold reloaded changes
+	attributes := map[string]string{}
```
I know it wasn't before, but doesn't this need to be `make(map[string]string)`?

mlaventure replied:
Happy to make the changes. Does it make a difference though?

cpuguy83 commented on daemon/daemon.go:
```diff
@@ -785,10 +789,24 @@ func (daemon *Daemon) initDiscovery(config *Config) error {
+	// used to hold reloaded changes
+	attributes := map[string]string{}
+
+	// We need defer here to ensure the lock is released as
```
I'm not sure I understand this comment since there is nothing in the defer about locking.

mlaventure replied:
If we put this defer after `defer daemon.configStore.reloadLock.Unlock()` the lock would still be held when calling `daemon.LogDaemonEventWithAttributes()` which would cause a deadlock

cpuguy83 commented:
LGTM

mlaventure committed:
Update containerd and runc vendoring

containerd: 860f3a94940894ac0a106eff4bd1616a67407ee2
runc: 85873d917e86676e44ccb80719fcb47a794676a1
runtime-specs: v1.0.0-rc1

Signed-off-by: Kenfe-Mickael Laventure <mickael.laventure@gmail.com>

mlaventure commented:
I seem to be hitting the issue on `DockerSuite.TestRunResolvconfUpdate` all the time on CI, never seen once on my host.

Also, those tests are failing, but I can't reproduce the failure locally either:
- experimental: `DockerExternalGraphdriverSuite.TestExternalGraphDriver`
- gccgo: `DockerSuite.TestPortExposeHostBinding`

Couldn't find an issue for those 2, known flaky @thaJeztah ?

Since those test are not touching the new added code, I don't see how this PR would affect them.

thaJeztah commented:
Looking..

this one failed on both experimental and userns here;

https://jenkins.dockerproject.org/job/Docker-PRs-experimental/19802/console

```
20:53:34 FAIL: docker_cli_run_test.go:1434: DockerSuite.TestRunResolvconfUpdate
20:53:34 
20:53:34 docker_cli_run_test.go:1514:
20:53:34     c.Fatalf("Container's resolv.conf should not have been updated with host resolv.conf: %q", string(containerResolv))
20:53:34 ... Error: Container's resolv.conf should not have been updated with host resolv.conf: "# Dynamic resolv.conf(5) file for glibc resolver(3) generated by resolvconf(8)\n#     DO NOT EDIT THIS FILE BY HAND -- YOUR CHANGES WILL BE OVERWRITTEN\nnameserver 172.31.0.2\nsearch us-west-2.compute.internal\n"
```

https://jenkins.dockerproject.org/job/Docker-PRs-userns/10761/console

```
20:39:21 FAIL: docker_cli_run_test.go:1434: DockerSuite.TestRunResolvconfUpdate
20:39:21 
20:39:21 docker_cli_run_test.go:1514:
20:39:21     c.Fatalf("Container's resolv.conf should not have been updated with host resolv.conf: %q", string(containerResolv))
20:39:21 ... Error: Container's resolv.conf should not have been updated with host resolv.conf: "# Dynamic resolv.conf(5) file for glibc resolver(3) generated by resolvconf(8)\n#     DO NOT EDIT THIS FILE BY HAND -- YOUR CHANGES WILL BE OVERWRITTEN\nnameserver 172.31.0.2\nsearch us-west-2.compute.internal\n"
```

Not sure about the other two actually; I can restart CI once to see if it's just really a flaky error

mlaventure commented:
Updated with new containerd/runc vendoring

icecrime commented:
@mlaventure Doesn't build :crying_cat_face: 

```
13:32:21 # github.com/docker/docker/daemon
13:32:21 daemon/update_linux.go:12: cannot use uint32(resources.BlkioWeight) (type uint32) as type uint64 in assignment
13:32:21 daemon/update_linux.go:13: cannot use uint32(resources.CPUShares) (type uint32) as type uint64 in assignment
13:32:21 daemon/update_linux.go:14: cannot use uint32(resources.CPUPeriod) (type uint32) as type uint64 in assignment
13:32:21 daemon/update_linux.go:15: cannot use uint32(resources.CPUQuota) (type uint32) as type uint64 in assignment
13:32:21 daemon/update_linux.go:18: cannot use uint32(resources.Memory) (type uint32) as type uint64 in assignment
13:32:21 daemon/update_linux.go:20: cannot use uint32(resources.MemorySwap) (type uint32) as type uint64 in assignment
13:32:21 daemon/update_linux.go:22: cannot use uint32(resources.MemoryReservation) (type uint32) as type uint64 in assignment
13:32:21 daemon/update_linux.go:23: cannot use uint32(resources.KernelMemory) (type uint32) as type uint64 in assignment
13:32:22 Build step 'Execute shell' marked build as failure```

icecrime commented:
LGTM (assuming green).

cpuguy83 commented:
LGTM

cpuguy83 commented:
Moving to docs review, we can kick back if there's a test failure.

icecrime commented on docs/reference/commandline/create.md:
```diff
@@ -78,6 +78,7 @@ Creates a new container.
       --privileged                  Give extended privileges to this container
       --read-only                   Mount the container's root filesystem as read only
       --restart="no"                Restart policy (no, on-failure[:max-retry], always, unless-stopped)
+      --runtime=""                  Name of the runtime to be used for that container (available runtimes are listed in `docker info`)
```
I would remove the `(available runtimes are listed in \`docker info\`)` part.

icecrime commented on docs/reference/commandline/dockerd.md:
```diff
@@ -1014,7 +1015,19 @@ This is a full example of the allowed configuration options in the file:
 	"insecure-registries": [],
-	"disable-legacy-registry": false
+	"disable-legacy-registry": false,
+	"default-runtime": "foo",
```
s/foo/runc

icecrime commented on docs/reference/commandline/dockerd.md:
```diff
@@ -1014,7 +1015,19 @@ This is a full example of the allowed configuration options in the file:
+	"disable-legacy-registry": false,
+	"default-runtime": "foo",
+	"runtimes": {
+		"foo": {
```
s/foo/runc

icecrime commented on docs/reference/commandline/dockerd.md:
```diff
@@ -1014,7 +1015,19 @@ This is a full example of the allowed configuration options in the file:
+		"foo": {
+			"path": "runc"
+		},
+		"bar": {
```
s/bar/custom?

icecrime commented on docs/reference/commandline/run.md:
```diff
@@ -89,6 +89,7 @@ parent = "smn_cli"
       --read-only                   Mount the container's root filesystem as read only
       --restart="no"                Restart policy (no, on-failure[:max-retry], always, unless-stopped)
       --rm                          Automatically remove the container when it exits
+      --runtime=""                  Name of the runtime to be used for that container (available runtimes are listed in `docker info`)
```
Same remark as above.

icecrime commented on docs/reference/commandline/dockerd.md:
```diff
@@ -60,6 +60,7 @@ weight = -1
       -p, --pidfile="/var/run/docker.pid"    Path to use for daemon PID file
       --raw-logs                             Full timestamps without ANSI coloring
       --registry-mirror=[]                   Preferred Docker registry mirror
+      --add-runtime=[]                       Register an additional OCI compatible runtime
```
It might be worth adding a subsection [here](https://github.com/mlaventure/docker/blob/90a306b1cf28c1bebb4990903225e8ad4f4170d1/docs/reference/commandline/dockerd.md#docker-runtime-execution-options) to explain how this can be used, and give concrete examples. From the documentation here, I have for example no way to tell what's the right syntax for the `--add-runtime` flag?

mlaventure commented:
Updated documentation with @icecrime suggestions

icecrime commented:
Docs LGTM (cc @thaJeztah).

mlaventure committed:
Fix race in DockerExternalGraphdriverSuite.TestExternalGraphDriver

Signed-off-by: Kenfe-Mickael Laventure <mickael.laventure@gmail.com>

mlaventure committed:
Vendor in new runc binary with userns fix

Signed-off-by: Kenfe-Mickael Laventure <mickael.laventure@gmail.com>

icecrime commented:
Still one failure in userns :grimacing: 

```
20:07:52 FAIL: docker_cli_run_test.go:1434: DockerSuite.TestRunResolvconfUpdate
20:07:52 
20:07:52 docker_cli_run_test.go:1514:
20:07:52     c.Fatalf("Container's resolv.conf should not have been updated with host resolv.conf: %q", string(containerResolv))
20:07:52 ... Error: Container's resolv.conf should not have been updated with host resolv.conf: "# Dynamic resolv.conf(5) file for glibc resolver(3) generated by resolvconf(8)\n#     DO NOT EDIT THIS FILE BY HAND -- YOUR CHANGES WILL BE OVERWRITTEN\nnameserver 172.31.0.2\nsearch us-west-2.compute.internal\n"
```

mlaventure committed:
Skip TestRunResolvconfUpdate as it is unstable

Signed-off-by: Kenfe-Mickael Laventure <mickael.laventure@gmail.com>

SvenDowideit commented:
I can't see anything describing what a `runtime` _is_, or where someone might find one, or write one.

needs more docs - otherwise, it looks a lot like you're creating an option that can't be used.

mlaventure commented:
@SvenDowideit https://github.com/mlaventure/docker/blob/610d9e12eeabcee9b53f453963d3c8d731bef5a8/docs/reference/commandline/dockerd.md#docker-runtime-execution-options ?

thaJeztah commented:
LGTM

thaJeztah commented:
retest this please

icecrime commented:
- Documentation is broken by #23492 
- gccgo is temporarily skipped by #23526
- windowsTP5 is a timeout

WDYT @crosbymichael @thaJeztah? Are we ok to merge?

crosbymichael commented:
LGTM

lousscott commented:
Take down
